### DISCUSSION TITLE
How to make X-Ray traces clickable from CloudWatch logs insights?

### DISCUSSION THREAD

**Discussion by softwareengineerprogrammer** [aws-powertools/powertools-lambda-python#5741] [Q&A] [answered]
I recently integrated powertools logging into my Python app, and love that it automatically instruments `xray_trace_id` in my cloudwatch logs. However, it would be significantly more useful to be able to click the trace ID in order to view the full trace, as CW dashboards supports when the trace ID is instrumented as `@xrayTraceId`, and is done automatically for `REPORT` events.

Examples - `xray_trace_id` not clickable in CW dashboard logs widget for an event logged by powertools:

<img width="684" alt="xray-trace-id-not-clickable" src="https://github.com/user-attachments/assets/3a322309-7685-459f-ab8a-ab35165b2cd9" />

`@xrayTraceId` clickable for `REPORT` event:

<img width="680" alt="xray-trace-id-clickable" src="https://github.com/user-attachments/assets/b57202d7-5bdf-4d3a-adba-cf7e9c2bfd45" />

What do I need to do to achieve clickable xray traces for all events as shown in the second image?

---

Edit: In the meantime, I created a workaround in the form of a [Tampermonkey](https://tampermonkey.net) userscript that adds a link for `xray_trace_id` entries. [install](https://softwareengineerprogrammer.github.io/userscripts/xray-trace-clickification.user.js) | [source](https://github.com/softwareengineerprogrammer/softwareengineerprogrammer.github.io/blob/main/userscripts/xray-trace-clickification.user.js)

<img width="677" alt="xray-trace-clickification-userscript-screenshot" src="https://github.com/user-attachments/assets/dfb26080-160e-4bfd-a14d-eb2c4b041a9b" />

**Comment by leandrodamascena** [comment, marked as answer]
Hey @softwareengineerprogrammer! If you want that CW Logs Dashboard render the `TracerId` with a link to Tracer, then you need to log the field `X-Amzn-Trace-Id` as part of your log line. In this way CW will detect this field and render this as `@xRayTracerId`. With Powertools you have some ways to do that.

1 - Adding this as an extra key. You must take care to get `_X_AMZN_TRACE_ID` from the env, it changes on every Lambda run and it doesn't matter if this is a new container or not, so put this inside the Lambda handler.

```python
import os
from aws_lambda_powertools import Logger

logger = Logger()

def lambda_handler(event: dict, context: LambdaContext) -> str:
	mylog = {
	    "X-Amzn-Trace-Id": os.getenv("_X_AMZN_TRACE_ID")
	}
	
	logger.info("Message",**mylog)
```

2 - Creating a `CustomFormatter` and injecting this key

```python
import os
from aws_lambda_powertools import Logger
from aws_lambda_powertools.logging.formatter import LambdaPowertoolsFormatter
from aws_lambda_powertools.logging.types import LogRecord


class CustomFormatter(LambdaPowertoolsFormatter):
    def serialize(self, log: LogRecord) -> str:
        """Serialize final structured log dict to JSON str"""
        log["X-Amzn-Trace-Id"] = os.getenv("_X_AMZN_TRACE_ID")
        return self.json_serializer(log)


logger = Logger(service="payment", logger_formatter=CustomFormatter())
logger.info("hello")
```

3 - I do no advice you to use [append_keys](https://docs.powertools.aws.dev/lambda/python/develop/core/logger/#append_keys-method) method because you need to get a new value on every execution.

Please ping me if you have any additional question.

**Reply by softwareengineerprogrammer** [reply]
Thank you @leandrodamascena! I haven't had time to incorporate your answer into my code yet but I really appreciate it.
